### PR TITLE
Add more generic ProductIcon component that is going to replace PlanIcon

### PR DESCRIPTION
**Pull request by delawski**
This PR introduces a `<ProductIcon />` - a more generic version of `<PlanIcon />` component that supports not only plans but also products icons. It's needed in the new `<PurchasesListing />` component to support Jetpack Backup (#37727).

Screen Shot on 2019-11-20 at 13-21-55

The current plan is to replace all occurrences of `<PlanIcon />` with this new component in the subsequent PR.

#### Changes proposed in this Pull Request

* Introduce a `<ProductIcon />` that's based and is meant to replace the `<PlanIcon />` component

#### Testing instructions

* Go to http://calypso.localhost:3000/devdocs/design/product-icon
* Confirm you see all sorts of icons (plans, Jetpack plans and Jetpack products)

Fixes n/a

### CODE REVIEW

File: client/components/product-icon/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+- PLAN_FREE,
+- PLAN_PREMIUM,
+- PLAN_BUSINESS,
+- PLAN_JETPACK_FREE,
+- PLAN_JETPACK_BUSINESS,
+- PLAN_JETPACK_BUSINESS_MONTHLY,
+- PLAN_JETPACK_PREMIUM,
+- PLAN_JETPACK_PREMIUM_MONTHLY,
+- PLAN_PERSONAL,
```
Comment (enejb) [inline client/components/product-icon/README.md]: Lets move this higher up.

File: client/components/product-icon/index.jsx [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		}
+
+		if ( isBusinessPlan( product ) ) {
+			return this.getPlanIcon( 'business' );
+		}
+
+		if ( isEcommercePlan( product ) ) {
+			return this.getPlanIcon( 'ecommerce' );
+		}
+
```
Comment (enejb) [inline client/components/product-icon/index.jsx]: My only is that folks could be on some sort of old plan that is currently doesn't seem to be supported here just yet.

**Comment by enejb** [comment]
Should we add support for other products such as domain?

File: client/components/product-icon/index.jsx (hunk 2 above)
Comment (delawski) [reply client/components/product-icon/index.jsx]: I think that if it was the case, this kind of bug would have been already fixed since this code is [migrated from the existing `PlanIcon` component](https://github.com/Automattic/wp-calypso/blob/master/client/components/plans/plan-icon/index.jsx#L45-L70).

Also, there are no more SVG icons available to us that we could use here. We've got just [6 icons for regular plans](https://github.com/Automattic/wp-calypso/tree/master/static/images/plans/wpcom) and [4 icons for Jetpack plans](https://github.com/Automattic/wp-calypso/tree/master/static/images/plans/jetpack).

**Comment by delawski** [comment]
> Should we add support for other products such as domain?

As per [my earlier comment](https://github.com/Automattic/wp-calypso/pull/37760#discussion_r348468011), I don't think we should worry about other products. AFAIK we are not making use of other icons anyway.

File: client/components/product-icon/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+ProductIcon Component
+=============
+
+ProductIcon component is a React component to display plans or products icons in SVG format.
```
Comment (tyxla) [inline client/components/product-icon/README.md]:
```suggestion
ProductIcon component is a React component to display plan or product icons in SVG format.
```

File: client/components/product-icon/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+We have four different plans icons:
+- Free
+- Personal
+- Premium
+- Business
```
Comment (tyxla) [inline client/components/product-icon/README.md]: Minor: should we specify that Business is AKA "Professional"?
```suggestion
- Business (Professional)
```

File: client/components/product-icon/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
+ProductIcon Component
+=============
+
+ProductIcon component is a React component to display plans or products icons in SVG format.
+
+We have four different plans icons:
```
Comment (tyxla) [inline client/components/product-icon/README.md]: Nit: It's generally more flexible if we just say that we have "the following" without specifying a number.

```suggestion
We have the following different plans icons:
```

File: client/components/product-icon/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+- Free
+- Personal
+- Premium
+- Business
+
+Jetpack plans also use these plans icons.
+
+There is also one product icon:
```
Comment (tyxla) [inline client/components/product-icon/README.md]: Not specifying the number keeps this readme a bit more flexible:

```suggestion
There are also the following product icons:
```

File: client/components/product-icon/README.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+Plan constant from `lib/plans/constants` or product constant from `lib/products-values/constants`. Can be one of:
+
+- PLAN_FREE,
+- PLAN_PERSONAL,
+- PLAN_PREMIUM,
+- PLAN_BUSINESS,
+- PLAN_JETPACK_FREE,
+- PLAN_JETPACK_BUSINESS,
+- PLAN_JETPACK_BUSINESS_MONTHLY,
```
Comment (tyxla) [inline client/components/product-icon/README.md]: I see we support monthly variants of Jetpack plans and Jetpack products, but we don't support 2-year plans for .com sites. Is that intentional?

File: client/components/product-icon/index.jsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
+/**
+ * External dependencies
+ */
+
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]:
```suggestion
```

File: client/components/product-icon/README.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+## Props
+
+### `product`
+
+Plan constant from `lib/plans/constants` or product constant from `lib/products-values/constants`. Can be one of:
+
+- PLAN_FREE,
+- PLAN_PERSONAL,
+- PLAN_PREMIUM,
+- PLAN_BUSINESS,
```
Comment (tyxla) [inline client/components/product-icon/README.md]: Guess we don't have icons for WP.com blogger or ecommerce plans, is that intentional?

File: client/components/product-icon/index.jsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		const { product, className } = this.props;
+
+		const plan = getPlan( product );
+		const planTitle = plan.getTitle();
+		const planClass = getPlanClass( product );
+		const isJetpack = isJetpackPlanSlug( product );
+
+		return (
+			<img
+				src={ `/calypso/images/plans/${ isJetpack ? 'jetpack' : 'wpcom' }/plan-${ iconName }.svg` }
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]: I think this like would be a little more readable if we move the ternary in its own variable above the `return`.

File: client/components/product-icon/index.jsx [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+				className={ classNames(
+					'product-icon',
+					`product-icon__${ iconName }`,
+					planClass,
+					className
+				) }
+				alt={ planTitle }
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]: In terms of accessibility, these images are with purely decorative purpose, so do they really need an `alt`? Also, should we add `role="presentation"`?

File: client/components/product-icon/index.jsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+		return (
+			<img
+				src={ `/calypso/images/products/jetpack/product-${ iconName }.svg` }
+				className={ classNames(
+					'product-icon',
+					`product-icon__${ iconName }`,
+					productClass,
+					className
+				) }
+				alt={ productTitle }
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]: In terms of accessibility, these images are with purely decorative purpose, so do they really need an `alt`? Also, should we add `role="presentation"`?

File: client/components/product-icon/index.jsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+	isJetpackPlanSlug,
+} from 'lib/products-values';
+
+/**
+ * Style dependencies
+ */
+import './style.scss';
+
+export default class ProductIcon extends Component {
+	getPlanIcon( iconName ) {
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]: Is there a reason why we're not using `<PlanIcon />` here?

File: client/components/product-icon/test/test.jsx [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+		/>
+	),
+	numberFormat: x => x,
+	translate: x => x,
+} ) );
+
+/**
+ * External dependencies
+ */
+import { assert } from 'chai';
```
Comment (tyxla) [inline client/components/product-icon/test/test.jsx]: Any good reason to not use `jest`'s built-in assertions (`expect`) and use `chai` instead?

File: client/components/product-icon/index.jsx [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+		const productTitle = getProductShortName( product );
+		const productClass = getProductClass( product );
+
+		return (
+			<img
+				src={ `/calypso/images/products/jetpack/product-${ iconName }.svg` }
+				className={ classNames(
+					'product-icon',
+					`product-icon__${ iconName }`,
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]: Not sure if this is material for this PR, but it felt odd to have the same result for 2 different products:



Should these differ in color perhaps?

File: client/components/product-icon/index.jsx [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+					productClass,
+					className
+				) }
+				alt={ productTitle }
+			/>
+		);
+	}
+
+	render() {
```
Comment (tyxla) [inline client/components/product-icon/index.jsx]: This way of having a separate condition for each one isn't very flexible, I wonder if we could have some sort of mapping that would make this method much smaller. Not a blocker, and I recognize that the `PlanIcon` does the same, so feel free to skip for now if you like.

File: client/components/product-icon/test/test.jsx [hunk 17]
```diff
@@ -0,0 +1,104 @@
+jest.mock( 'lib/abtest', () => ( {
```
Comment (tyxla) [inline client/components/product-icon/test/test.jsx]: Why is that necessary?

Comment (tyxla) [inline client/components/product-icon/test/test.jsx]: It feels odd to have those at the top of the file. I'd expect them to be in a `beforeEach` in the corresponding `describe` call.

File: client/lib/plans/index.js [hunk 18]
```diff
@@ -30,6 +30,10 @@ export function getPlans() {
 	return PLANS_LIST;
 }
 
+export function getPlansKeys() {
```
Comment (tyxla) [inline client/lib/plans/index.js]: Any more appropriate term than `key` here? Plan slugs?

File: client/lib/products-values/index.js [hunk 19]
```diff
@@ -359,6 +370,22 @@ export function getDomain( product ) {
 	return product.meta;
 }
 
+export function getProductsKeys() {
```
Comment (tyxla) [inline client/lib/products-values/index.js]: Any more appropriate term than `key` here? Product slugs?

Or maybe we can directly use the constant where necessary? Although I don't mind having consistency with what we're doing for plans.

File: client/lib/products-values/index.js [hunk 20]
```diff
@@ -359,6 +370,22 @@ export function getDomain( product ) {
...
+export function getProductClass( productSlug ) {
+	if ( isJetpackBackupSlug( productSlug ) ) {
+		return 'is-jetpack-backup';
+	}
+
+	return '';
+}
+
+export function getProductShortName( productSlug ) {
+	return PRODUCT_SHORT_NAMES?.[ productSlug ] ?? '';
```
Comment (tyxla) [inline client/lib/products-values/index.js]: Is `null` a better return type than an empty string in this case?

File: client/lib/products-values/index.js [hunk 21]
```diff
@@ -168,11 +168,15 @@ export function isEnterprise( product ) {
 	return product.product_slug === PLAN_WPCOM_ENTERPRISE;
 }
 
+export function isJetpackPlanSlug( productSlug ) {
```
Comment (tyxla) [inline client/lib/products-values/index.js]: Would be great if we start adding tests for all those new helpers.

File: static/images/products/jetpack/product-backup.svg [hunk 22]
```diff
@@ -0,0 +1,10 @@
+<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 64 64">
```
Comment (tyxla) [inline static/images/products/jetpack/product-backup.svg]: Can we minify the svg?

File: client/components/product-icon/index.jsx (hunk 13 above)
Comment (delawski) [reply client/components/product-icon/index.jsx]: I'm planning on eventually replacing `PlanIcon` with `ProductIcon` globally. This component is an enhanced version of `PlanIcon`.

The reason why those enhancement did not land inside `PlanIcon` is that we want to change its name to something more generic.

File: client/components/product-icon/README.md (hunk 9 above)
Comment (delawski) [reply client/components/product-icon/README.md]: Well, actually we have icons for both. I overlooked this error when moving the code over from `PlanIcon` component.

File: client/components/product-icon/README.md (hunk 7 above)
Comment (delawski) [reply client/components/product-icon/README.md]: According to tests which are passing, we support 2-year plans. I'm going to update the readme file.

File: client/components/product-icon/index.jsx (hunk 11 above)
Comment (delawski) [reply client/components/product-icon/index.jsx]: In `PlanIcon` there's an `eslint-disable jsx-a11y/alt-text` rule around `img` so I thought of adding `alt` in order to get rid of this. But what you're saying makes sense to me too. Since it's a presentational element, we can go with `role="presentation"` instead.

File: client/components/product-icon/index.jsx (hunk 15 above)
Comment (delawski) [reply client/components/product-icon/index.jsx]: @eeeeevon13 What do you think about @tyxla's suggestion?

File: client/components/product-icon/index.jsx (hunk 16 above)
Comment (delawski) [reply client/components/product-icon/index.jsx]: I'm going to think about it 👍

File: client/components/product-icon/test/test.jsx (hunk 17 above)
Comment (delawski) [reply client/components/product-icon/test/test.jsx]: I copied the whole test file from `PlanIcon` and haven't checked it. Will double check 👍

File: client/components/product-icon/test/test.jsx (hunk 14 above)
Comment (delawski) [reply client/components/product-icon/test/test.jsx]: I copied the whole test file from `PlanIcon` and haven't changed it at all. It's probably a legacy approach. I'm going to change this.

File: client/lib/products-values/index.js (hunk 19 above)
Comment (delawski) [reply client/lib/products-values/index.js]: I'd prefer to use a getter instead of accessing the constant directly.

File: static/images/products/jetpack/product-backup.svg (hunk 22 above)
Comment (delawski) [reply static/images/products/jetpack/product-backup.svg]: I've run it through SVGOMG but I can see that the whitespace has not been removed. I wonder if removing the whitespace is worth it anyway.

File: client/components/product-icon/test/test.jsx (hunk 14 above)
Comment (tyxla) [reply client/components/product-icon/test/test.jsx]: Yeah, some of those tests were never migrated to Jest. But it's recommended that we use it for any new tests.

File: client/lib/products-values/index.js (hunk 19 above)
Comment (tyxla) [reply client/lib/products-values/index.js]: Sounds good to me 👍

File: static/images/products/jetpack/product-backup.svg (hunk 22 above)
Comment (tyxla) [reply static/images/products/jetpack/product-backup.svg]: Well, we usually remove the whitespace, as it's really unnecessary burden for the client to download, and we don't currently minify the svgs out of the box.

File: client/components/product-icon/index.jsx (hunk 11 above)
Comment (delawski) [reply client/components/product-icon/index.jsx]: ESLint wasn't happy about having a `role="presentation"` instead of an `alt` so I've reverted to having the `return` statement between `eslint-disable jsx-a11y/alt-text` rules.

**Comment by delawski** [comment]
@tyxla I've addressed your feedback. It's now ready for another round of code review.
